src/metric/elementwise_metric.cu

```diff
@@ -120,8 +116,9 @@ class ElementWiseMetricsReduction {
       for (GPUSet::GpuIdType id = *devices.begin(); id < *devices.end(); ++id) {
         dh::safe_cuda(cudaSetDevice(id));
         size_t index = devices.Index(id);
+        dh::CubMemory allocator;
         res_per_device.at(index) =
-            DeviceReduceMetrics(id, index, weights, labels, preds);
+            DeviceReduceMetrics(id, &allocator, weights, labels, preds);
```

Comment:
i did verify this by training on a large dataset, and did not notice any meaningful difference.

Reply:
In general, I don't want memory allocations occurring on a per iteration basis. Another way of solving this would be to manually call cudaSetDevice and allocator.at(i).Free() in the destructor of the metric class. I am about to overhaul the memory allocators using a global cub caching allocator in #4554. This will make the CubMemory allocator class redundant soon.

Reply:
please check now. i was hoping to not add additional state, as i did not see any performance regression. i have now pre-created the allocators *exactly* once to prevent recycling the allocators during implicit resizes.

Reply:
Those resizes are a no op anyway assuming the number of GPUs remains constant. But I think what you have is fine.

Reply:
yes, practically, they may not be. but, if the device list passed to the function (which is a function of the label size) keeps increasing (it was 1 when eval was invoked, and then 2 the next time etc. - capped to the number of devices), then it may trigger a resize, which would then lead to destruction of the previous cubmemory instances, thus causing scenarios this pr is trying to prevent.